lesson-003-web-server-graceful-shutdown/README.md

```diff
@@ -0,0 +1,104 @@
+
+```
+func say(s string) {
+	for i := 0; i < 5; i++ {
+		time.Sleep(100 * time.Millisecond)
+		fmt.Println(s)
+	}
+}
+
+func main() {
+	go say("k8s")
+	say("contribute")
+}
+```
+
+Calling `go printMyString()` is where we start a new `goroutine`, calling our already defined `printMyString` function, while the `main` function runs in its own `goroutine` (called the _main goroutine_)
+
+
+### Channels
+
+Channels are a concurrency synchronization technique we can use in Go.
+
+We can create a channel using the keyword `chan`, and we can transport data of only one data type.
+
+From the [official documention](https://tour.golang.org/concurrency/2) we can see how data can be sent and received from the channel.
+
+```
+ch := make(chan int)
+
+ch <- v    // Send v to channel ch.
+v := <-ch  // Receive from ch, and
+           // assign value to v.
+```
+
+**_Unbuffered_ vs _buffered_ channels**
+
+Channels are created **blocking by default** (or _unbuffered_)
+
+Unbuffered channels are those which both goroutines (one sending data to the buffer, another waiting to receive the resource) require to be ready to exchange some data. If a goroutine is trying to send a resource to an unbuffered channel and there is no goroutine waiting to receive that resource, the channel will lock the sender and make it wait.
+And viceversa.
+
+On the other hand, buffered channels have capacity and they are able to keep a number of resources.
+The only times buffered channels will lock goroutines are: when a sender tries to send a resource and the channel is full or when a goroutine tries to get a resource and the channel is empty.
+
+We can use this in our benefit and block our server from being closed until all pending requests have been served when we receive a specific signal.
+
+
+### Signals
+
+Signals are software interrupts sent to a program to indicate that an event has happened.
+
+In this example we will take care of two specific signals: `os.Interrupt` and `syscall.SIGTERM`:
+
+- `os.Interrupt` this is tipically the signal sent when we type Control-C, which normally causes the program to exit.
+- `syscall.SIGTERM` is usually sent when you want to give the process an opportunity to clean up before termination.
+
+
+## Lesson content
+
+After getting a quick intro to `channels` and `goroutines`, we will dive into our lesson.
+
+As we initially stated, the current implementation is based on our server from [lesson 000](../lesson-000-web-server/).
+
+
+To achieve _graceful shutdown_ we added the following changes:
+
+- created a `sigint` channel; we will use it to notify our goroutine we have received a signal: `os.Interrupt` or `syscall.SIGTERM` in this case
+- with `server.SetKeepAlivesEnabled(false)` we set the server to not keep alive any connection (which in fact is the desired effect of having a gracefull shutdown behavior)
+- create the `done` channel; this one will be used to let the main goroutine we have finished the graceful shutdown.
```

Comment:
useful to mention that adding this in main allows us to run code only after the app is shut down - if needed.